linked_list/linked_list.py

```diff
@@ -13,67 +12,157 @@ def __init__(self):
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        current_node = self.head
+        current_index = 0
+
+        if current_node is None:
+            return None
+        
+        while current_node is not None:
+            if current_index == index:
+                return current_node.value
+            else:
+                current_node = current_node.next
+                current_index+=1
+        return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        current_node = self.head
+
+        if current_node is None:
+            return None
+        
+        while True:
+            if current_node.next is None:
+                return current_node.value
+            else:
+                current_node = current_node.next
+
+        
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        new_node = Node(value)
+        current_node = self.head
+
+        if current_node is None:
+            self.head = new_node
+            return
+
+        # What do I want while loop to do? I want while loop to find the last node 
+
+        while current_node is not None:
+            if current_node.next is None:
+                break
+            current_node = current_node.next 
+
+        # When last node is found, I want to append new_node to it.
+        current_node.next = new_node
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        current_node = self.head
+        if current_node is None:
+            return None
+
+        max_value = current_node.value
+
+        while current_node.next:
+            if current_node.value > max_value:
+                max_value = current_node.value
+            current_node = current_node.next
+
+        if current_node.value > max_value:
+            max_value = current_node.value
+              
+        return max_value
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def delete(self, value):
-        pass
+        current_node = self.head
+        previous_node = None
+
+        if current_node is None:
+            return None
+
+        # Case: value to be deleted is the first value in the linked list
+        if current_node.value == value:
+            self.head = current_node.next
+            current_node = None

```

Comment:
We don't need to re-assign this after we've skipped over it and connected it to the head node

```suggestion
```